src/upload.rs

```diff
@@ -74,48 +79,56 @@ where
     }
 }
 
-impl Decodable for Keyword {
-    fn decode<D: Decoder>(d: &mut D) -> Result<Keyword, D::Error> {
-        let s = d.read_str()?;
+impl<'de> Deserialize<'de> for Keyword {
+    fn deserialize<D: Deserializer<'de>>(d: D) -> Result<Keyword, D::Error> {
+        let s = String::deserialize(d)?;
         if !CrateKeyword::valid_name(&s) {
-            return Err(d.error(&format!("invalid keyword specified: {}", s)));
+            let value = de::Unexpected::Str(&s);
+            let expected = "a valid keyword specifier";
+            Err(de::Error::invalid_value(value, &expected))
+        } else {
+            Ok(Keyword(s))
         }
-        Ok(Keyword(s))
     }
 }
 
-impl Decodable for Category {
-    fn decode<D: Decoder>(d: &mut D) -> Result<Category, D::Error> {
-        d.read_str().map(Category)
-    }
-}
-
-impl Decodable for Feature {
-    fn decode<D: Decoder>(d: &mut D) -> Result<Feature, D::Error> {
-        let s = d.read_str()?;
+impl<'de> Deserialize<'de> for Feature {
+    fn deserialize<D: Deserializer<'de>>(d: D) -> Result<Feature, D::Error> {
+        let s = String::deserialize(d)?;
         if !Crate::valid_feature_name(&s) {
-            return Err(d.error(&format!("invalid feature name specified: {}", s)));
+            let value = de::Unexpected::Str(&s);
+            let expected = "a valid feature name";
+            Err(de::Error::invalid_value(value, &expected))
+        } else {
+            Ok(Feature(s))
         }
-        Ok(Feature(s))
     }
 }
 
-impl Decodable for CrateVersion {
-    fn decode<D: Decoder>(d: &mut D) -> Result<CrateVersion, D::Error> {
-        let s = d.read_str()?;
+impl<'de> Deserialize<'de> for CrateVersion {
+    fn deserialize<D: Deserializer<'de>>(d: D) -> Result<CrateVersion, D::Error> {
+        let s = String::deserialize(d)?;
         match semver::Version::parse(&s) {
             Ok(v) => Ok(CrateVersion(v)),
-            Err(..) => Err(d.error(&format!("invalid semver: {}", s))),
+            Err(..) => {
+                let value = de::Unexpected::Str(&s);
+                let expected = "a valid semver";
+                Err(de::Error::invalid_value(value, &expected))
```

Comment:
I tried to at first, but failed. Because of conduit, we must use semver 0.5, which doesn't have serde feature.